<a href="https://colab.research.google.com/github/IldikoW/CFG/blob/main/Homework1/HomeWork1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Question 1  
In no more than 500 words, explain the differences between a Jupyter Notebook (.ipynb
extension) and a Python file (.py extension). Focus on the advantages and disadvantages of
one or the other, particularly for data analysts/scientists. You are encouraged to use
examples and screenshots of the different scripts to support your arguments.

**Jupyter Notebooks (.ipynb) and Python files (.py) are two different formats , jupyter nootbook files ending with .ipynb and Python files end with .py**

# **Jupyter Notebooks (.ipynb):**

**Advantages:**

Interactivity: Jupyter Notebooks provide an interactive environment where code can be executed cell after cell. This is advantageous for data analysis tasks as it allows users to visualize intermediate results, making it easier to understand and debug code.
Rich Media Support: Jupyter Notebooks support the integration of rich media: images, plots, and HTML. This is particularly beneficial for data scientists who want to create and share documents that combine code, visualizations, and explanations in a single file.

Documentation and Exploration: Jupyter Notebooks allow for the inclusion of markdown cells, making the creation of documentation easy within the same document as the code. This is useful for data analysts who need to explain their thought process, document data cleaning steps, and share insights with others.

**Disadvantages:**

Version Control: The structure of .ipynb files makes it challenging to manage changes effectively. This can result in conflicts and difficulties in collaborative coding environments.

Execution Order Dependency: The order in which cells are executed can affect the notebook's output. This can lead to confusion, eit requires a clear understanding of the notebook's execution flow.

# Python Files (.py):

**Advantages:**

Modularity: Python files encourage a modular approach to coding, promoting the creation of functions and classes that can be reused in other scripts. This is beneficial for  collaboration.
Version Control: Python files are inherently more friendly to version control systems like Git. Changes are easier to track, making collaboration smoother and reducing the likelihood of conflicts.

**Disadvantages:**

Lack of Interactivity: Python scripts are typically executed from start to finish, lacking the interactivity offered by Jupyter Notebooks. This makes it more challenging to explore and visualize intermediate results during the development process.

Limited Rich Media Integration: While it is possible to include visualizations and comments in Python scripts, the integration of rich media is not as seamless as in Jupyter Notebooks. This may impact the ability to create comprehensive and visually appealing reports directly within the script.

Question 2  [10 points]
What is the difference between a Pandas DataFrame and a Pandas series. Show an
example of how you create each of them.

**A Pandas DataFrame is a two-dimensional, tabular data structure with labeled axes (rows and columns). It is similar to a spreadsheet or SQL table, where data is organized in rows and columns. Each column in a DataFrame can have a different data type, and you can perform various operations like indexing, selection, filtering, and merging on DataFrames.**

In [ ]:
# Creating a DataFrame from a dictionary
data = {
    'Name': ['John', 'Jane', 'Bob', 'Alice'],
    'Age': [25, 30, 22, 35],
    'City': ['New York', 'Los Angeles', 'Chicago', 'Houston']
}

df = pd.DataFrame(data)

# Display the DataFrame
print("Pandas DataFrame:")
print(df)

**A Pandas Series is a one-dimensional labeled array that can hold data of any type. It is like a single column or a single row of a DataFrame. Each element in a Series has a label (index), and you can perform various operations like indexing, slicing, and arithmetic operations on Series.**

In [ ]:
# Creating a Series from a list
ages = pd.Series([25, 30, 22, 35], name='Age')

# Display the Series
print("Pandas Series:")
print(ages)

Question 3  [10 points]
Starting from the argument that a Pandas DataFrame represents rectangular data, use the
internet and other resources to describe in no more than a few sentences  the difference
between rectangular and non-rectangular data.

**A Pandas DataFrame is like a table with rows and columns, and we call this rectangular data. It is organized and easy to work with, just like a spreadsheet or a database. Non-rectangular data, however, is a bit different.  It can be all kinds of shapes and structures, like nested data or graphs. Things like JSON files or XML documents fall into this category. Basically, rectangular data is like a well-organized table, while non-rectangular data can be more irregular and varied .**





Question 4  
Starting from the data visualisation usage from Session 2, give examples of when figures
could be:
a. Of use to the data scientist to identify patterns in the data/highlight the important
parts of a data set;
b. Tell a story and create business presentations.

#  Data scientist

 **Identifying Patterns and Trends:**

Example: Line charts, scatter plots, or heatmaps can help a data scientist identify patterns or trends in time-series data, correlations between variables, or spatial distributions.

**Data Distribution:**

Example: Histograms or kernel density plots can provide insights into the distribution of a variable, helping data scientists understand the underlying data structure.

# Business Presentations
**Forecasting and Planning:**

Example: Time-series forecasts visualized with line charts or area charts can assist in presenting future trends or predictions, aiding in business planning and strategy.

**Storytelling with Data:**

Example: A combination of line charts, bar graphs, and annotations can tell a compelling story about the growth or decline of key business metrics over time, helping to convey a narrative to stakeholders.

Question 5

In [2]:
import pandas as pd


In [9]:
df=titanic=pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

In [13]:
print(df.head())

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


In [14]:
print(df.describe())

         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


**I have printed out the first few rows and the describe so I would have an impression about the data and double check my answers.**


a. How many columns and rows does the data have?

In [5]:
print("Number of rows:", len(df))
print("Number of columns:", len(df.columns))

Number of rows: 891
Number of columns: 15


**The data has 15 columns and 891 rows.**

b, Get a sense of your data and find the min, max, count/mean.

In [15]:
print (df.count())



survived       891
pclass         891
sex            891
age            714
sibsp          891
parch          891
fare           891
embarked       889
class          891
who            891
adult_male     891
deck           203
embark_town    889
alive          891
alone          891
dtype: int64


This line shows how many data rows we have in each column, how big the data set is.

We can only calculate minimum , maximum and mean for numerical data for example age. We cannot calculate minimum, maximum and mean for a categorical data for example the column 'who' directly, without converting the data or boolean.
Accoring to the printout of the first few lines we can see which column has what type of data.

In [21]:
print (df['age'].max(),df['age'].min(),df['age'].mean())

80.0 0.42 29.69911764705882


**It show that the oldest passenger was 80 and the youngest was less then 1 year old. The average age of the passengers was around 30 years old.**

In [22]:
print (df['fare'].max(),df['fare'].min(),df['fare'].mean())

512.3292 0.0 32.204207968574636


**We can see that some people paid a lot of money (512) some nothing , they might be the people working onboard and the average fare was 32.**

In [33]:
print(df['sex'].value_counts().get('female', 0), df['sex'].value_counts().get('male',0),df['sex'].count())

314 577 891


**Here we can see that there were more male than female passengers onboard:almost twice as many.**

In [27]:
print(df['survived'].value_counts().get(0, 0), df['survived'].value_counts().get(1,0))
print(df['survived'].count())

549 342
891


**549 people did not survive the trip and 342 survived out of the total of 891 people.**

In [31]:
print(df['class'].value_counts().get('First', 0), df['class'].value_counts().get('Second',0),df['class'].value_counts().get('Third',0),df['class'].count())

216 184 491 891


**216 of the passngers travelled on first class, 184 travelled on second and 491 , almost half of the pessangers, on third class out of the total of 891.**

c, Give an overview (code an explanation) all missing values in the data.

In [32]:
missing_values = df.isnull().sum()
print("\nMissing values summary:")
print(missing_values)


Missing values summary:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


**The missing values come from the age coloumn an deck column. So in 177 cases we do not know the age of the passenger out of 891 passengers. In most cases, 688 , we do not know the deck data. In two cases out of 891 passengers we do not know the place of embarkment.**

d, Delete the rows where you do not have infomation about the age of the person.Then group the passengers in a 10 year age range.

In [ ]:
df = df.dropna(subset=['age'])

**This line deletes the rows where we do not have any information (NA)about the age of the person.**

In [ ]:
bins = [0, 10, 20, 30,40,50, 60, 70, 80]
labels = ['0-10', '11-20', '21-30','31-40','41-50','51-60','61-70','71-80']

df['Age_Group'] = pd.cut(df['age'], bins=bins, labels=labels, right=False)

In [ ]:
print(df['Age_Group'])

**The lines about create bins (categories) for the passengers according to their age in a 10 year age range. Then we print out the categories.**

e, For each age category created in d, find out how many passengers are female/male and how many travelled in each class.

In [ ]:
summary_table = df.groupby(['Age_Group', 'sex', 'class']).size().unstack(fill_value=0)

In [ ]:
print(summary_table)

class             First  Second  Third
Age_Group sex                         
0-10      female      1       8     21
          male        2       9     21
11-20     female     13       8     24
          male        5      10     42
21-30     female     16      25     31
          male       18      28    102
31-40     female     27      17     16
          male       23      31     53
41-50     female     13      10      9
          male       24       8     25
51-60     female     12       6      0
          male       15       9      6
61-70     female      3       0      1
          male       10       3      2
71-80     female      0       0      0
          male        3       1      2


**The table above shows the quantity of passengers by gender in each category accoring to  age groups and how many   travelled in each class.**